packages/ics721/src/testing/contract.rs

```diff
@@ -0,0 +1,386 @@
+        deps.querier = querier;
+        let env = mock_env();
+
+        let info = mock_info(NFT_ADDR, &[]);
+        let token_id = "1";
+        let sender = "ekez".to_string();
+        let msg = to_binary(&IbcOutgoingMsg {
+            receiver: "callum".to_string(),
+            channel_id: "channel-1".to_string(),
+            timeout: IbcTimeout::with_timestamp(Timestamp::from_seconds(42)),
+            memo: None,
+        })
+        .unwrap();
+
+        let res: cosmwasm_std::Response<_> = Ics721Contract::default()
+            .receive_nft(
+                deps.as_mut(),
+                env,
+                info,
+                TokenId::new(token_id),
+                sender.clone(),
+                msg,
+            )
+            .unwrap();
+        assert_eq!(res.messages.len(), 1);
+
+        let channel_id = "channel-1".to_string();
+        assert_eq!(
+            res.messages[0],
+            SubMsg::new(CosmosMsg::<Empty>::Ibc(IbcMsg::SendPacket {
+                channel_id: channel_id.clone(),
+                timeout: IbcTimeout::with_timestamp(Timestamp::from_seconds(42)),
+                data: to_binary(&NonFungibleTokenPacketData {
+                    class_id: ClassId::new(NFT_ADDR),
+                    class_uri: None,
+                    class_data: Some(
+                        to_binary(&ClassData {
+                            owner: Some(OWNER.to_string()),
+                            contract_info: expected_contract_info.clone(),
+                            name: "name".to_string(),
+                            symbol: "symbol".to_string(),
+                            num_tokens: 1,
+                        })
+                        .unwrap()
+                    ),
+                    token_data: None,
+                    token_ids: vec![TokenId::new(token_id)],
+                    token_uris: Some(vec!["https://moonphase.is/image.svg".to_string()]),
+                    sender,
+                    receiver: "callum".to_string(),
+                    memo: None,
+                })
+                .unwrap()
+            }))
+        );
+
+        // check outgoing classID and tokenID
+        let keys = OUTGOING_CLASS_TOKEN_TO_CHANNEL
+            .keys(deps.as_mut().storage, None, None, Order::Ascending)
+            .into_iter()
+            .collect::<StdResult<Vec<(String, String)>>>()
+            .unwrap();
+        assert_eq!(keys, [(NFT_ADDR.to_string(), token_id.to_string())]);
+
+        // check channel
+        let key = (
+            ClassId::new(keys[0].clone().0),
+            TokenId::new(keys[0].clone().1),
+        );
+        assert_eq!(
+            OUTGOING_CLASS_TOKEN_TO_CHANNEL
+                .load(deps.as_mut().storage, key)
+                .unwrap(),
+            channel_id
+        )
+    }
+    // test case: receive nft from old/v16 cw721-base
+    {
+        let mut querier = MockQuerier::default();
+        querier.update_wasm(mock_querier_v16);
```

Comment:
I think we can just have this in another named test something like `test_receive_nft_cw016`

Reply:
Before in other tests, I did that as well: putting all kinds of tests in various test functions. Later when adding, adjusting tests, I got lost. So I ended up grouping test functions covering a specific utility/contract function, and in there do tests on various cases for that specific function.